### PR TITLE
Conditional device includes, printf mask-out, basic programmatic serial access and BT boilerplate code

### PR DESCRIPTION
**Pull request by chaturatbs**
1. Conditional includes for devices - devices can be enabled/disabled by commenting out the #include statements.
2. #define mask-out for printf statements - printf statements can be disabled system wide with a #define

3. *EXPERIMENTAL* Hidden menu option '#' for repeated reads on AS7262 - Data piped through the serial connection and the RTT client python script can be used to process sensor readings on the serverside as well as display them on a GUI (see tools/gui). Can be enabled by un-commenting "#define FELIX" on the boot.c file. 
4. *EXPERIMENTAL* Java based GUI for displaying AS7262 sensor readings on a PC.
5. *EXPERIMENTAL* python RTT_client for reading data from over a serial connection with multiprocessing support for serial pass-through to the GUI, data logging and to pipe PC stdin to warp stdin 
6. *EXPERIMENTAL* Basic boilerplate code for bluetooth implementation. Doesnt compile yet. Need to reduce code size. can be disabled by commenting out "#include "btstack_main.h"

### CODE REVIEW

File: src/boot/ksdk1.1.0/devADXL362.h [hunk 1]
```diff
@@ -35,6 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
+#ifndef warp_devADXL362_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devADXL362.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devAMG8834.h [hunk 2]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devAMG8834_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devAMG8834.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devAS7262.c [hunk 3]
```diff
@@ -54,6 +57,9 @@
 extern volatile WarpI2CDeviceState	deviceAS7262State;
 extern volatile uint32_t		gWarpI2cBaudRateKbps;
 
+//Public functions for toggling LED
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devAS7262.c]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devAS7262.c [hunk 4]
```diff
@@ -99,47 +107,15 @@ readSensorRegisterAS7262(uint8_t deviceRegister)
...
-		return kWarpStatusDeviceCommunicationFailed;
-	}
-
-
 
 	/*
 	 *	See Page 8 to Page 11 of AS726X Design Considerations for writing to and reading from virtual registers.
 	 *	Write transaction writes the value of the virtual register one wants to read from to the WRITE register 0x01.
 	 */
+	#ifdef AS7262_DGBUG
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devAS7262.c]: Please fix. We can't merge these debugging statements into master

File: src/boot/ksdk1.1.0/devAS7262.c [hunk 5]
```diff
@@ -159,6 +135,10 @@ readSensorRegisterAS7262(uint8_t deviceRegister)
 	 *	Read transaction which reads from the READ register 0x02.
 	 *	The read transaction requires one to first write to the register address one wants to focus on and then read from that address.
 	 */
+	#ifdef AS7262_DGBUG
+	SEGGER_RTT_WriteString(0, "\t\t AS7262 Driver : Sending data - set register to read\r\n");
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devAS7262.c]: Please fix. Can't add debugging statements into master.

File: src/boot/ksdk1.1.0/devAS7262.c [hunk 6]
```diff
@@ -188,7 +172,39 @@ readSensorRegisterAS7262(uint8_t deviceRegister)
...
+
+		return kWarpStatusDeviceCommunicationFailed;
+	}
+
+	return kWarpStatusOK;
+}
+
+
+WarpStatus
+LEDonAS7262(void) {
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devAS7262.c]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devAS7262.c [hunk 7]
```diff
@@ -204,9 +220,67 @@ readSensorRegisterAS7262(uint8_t deviceRegister)
...
+							&slave /* The pointer to the I2C device information structure */,
+							cmdBuf_LEDON /* The pointer to the commands to be transferred */,
+							2 /* The length in bytes of the commands to be transferred */,
+							NULL /* The pointer to the data to be transferred */,
+							0 /* The length in bytes of the data to be transferred */,
+							500 /* timeout in milliseconds */);
+
+	if (returnValue == kStatus_I2C_Success)
+	{
+		//...
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devAS7262.c]: This regresses the version in master which checks the return value and returns an error. See the code you deleted.

File: src/boot/ksdk1.1.0/devAS7262.c [hunk 8]
```diff
@@ -204,9 +220,67 @@ readSensorRegisterAS7262(uint8_t deviceRegister)
...
+					NULL /* The pointer to the data to be transferred */,
+					0 /* The length in bytes of the data to be transferred */,
+					500 /* timeout in milliseconds */);
+
 	/*
 	 *	This turns off the LED after finish reading the data
 	*/
+
+	#ifdef AS7262_DGBUG
+	SEGGER_RTT_WriteString(0, "\t\t AS7262 Driver : Sending data - set data to write\r\n");
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devAS7262.c]: Please clean this up. Will not merge debugging statements into master.

File: src/boot/ksdk1.1.0/devAS7262.c [hunk 9]
```diff
@@ -215,7 +289,12 @@ readSensorRegisterAS7262(uint8_t deviceRegister)
 							NULL /* The pointer to the data to be transferred */,
 							0 /* The length in bytes of the data to be transferred */,
 							500 /* timeout in milliseconds */);
-	if (returnValue != kStatus_I2C_Success)
+
+	if (returnValue == kStatus_I2C_Success)
+	{
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devAS7262.c]: This regresses the behavior in master

File: src/boot/ksdk1.1.0/devAS7262.h [hunk 10]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devAS7262_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devAS7262.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devAS7262.h [hunk 11]
```diff
@@ -44,3 +46,6 @@ WarpStatus	readSensorSignalAS7262(WarpTypeMask signal,
 					WarpSignalAccuracy accuracy,
 					WarpSignalReliability reliability,
 					WarpSignalNoise noise);
+
+WarpStatus LEDonAS7262(void);
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devAS7262.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devAS7263.h [hunk 12]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devAS7263_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devAS7263.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devBME680.h [hunk 13]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devBME680_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devBME680.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devBMX055.h [hunk 14]
```diff
@@ -35,6 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
+#ifndef warp_devBMX055_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devBMX055.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devCCS811.h [hunk 15]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devCCS811_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devCCS811.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devHDC1000.h [hunk 16]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devHDC1000_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devHDC1000.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devL3GD20H.h [hunk 17]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devL3GD20H_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devL3GD20H.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devLPS25H.h [hunk 18]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devLPS25H_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devLPS25H.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devMAG3110.h [hunk 19]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devMAG3110_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devMAG3110.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devMMA8451Q.h [hunk 20]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devMMA8451Q_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devMMA8451Q.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devPAN1326.h [hunk 21]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devPAN1326_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devPAN1326.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devSI4705.h [hunk 22]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devSI4705_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devSI4705.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devSI7021.h [hunk 23]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devSI7021_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devSI7021.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/devTCS34725.h [hunk 24]
```diff
@@ -35,7 +35,9 @@
 	POSSIBILITY OF SUCH DAMAGE.
 */
 
-
+#ifndef warp_devTCS34725_ENABLE
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/devTCS34725.h]: Please see the coding conventions at https://github.com/physical-computation/Warp-firmware/blob/master/doc/README-CodingConventions.md

File: src/boot/ksdk1.1.0/felix/felix_main.c [hunk 25]
```diff
@@ -0,0 +1,20 @@
+#include "felix_main.h"
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/felix/felix_main.c]: This file should not be in the pull request.

File: src/boot/ksdk1.1.0/felix/felix_main.h [hunk 26]
```diff
@@ -0,0 +1,2 @@
+
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/felix/felix_main.h]: This file should not be in the pull request.

File: src/boot/ksdk1.1.0/gpio_pins.h [hunk 27]
```diff
@@ -8,31 +8,31 @@
 /*
```
Comment (phillipstanleymarbell) [inline src/boot/ksdk1.1.0/gpio_pins.h]: This file contains unnecessary diffs that do not make any actual change.

File: .gitmodules [hunk 28]
```diff
@@ -0,0 +1,4 @@
+[submodule "src/btstack"]
```
Comment (phillipstanleymarbell) [inline .gitmodules]: The submodule should be the one from Bluekitchen.

File: .gitattributes [hunk 29]
```diff
@@ -0,0 +1,2 @@
+* text=auto
```
Comment (phillipstanleymarbell) [inline .gitattributes]: This file should not be part of the pull request.

File: src/boot/ksdk1.1.0/devAMG8834.h (hunk 2 above)
Comment (chaturatbs) [reply src/boot/ksdk1.1.0/devAMG8834.h]: is "kWarp_devAMG8834_ENABLE" ok? or is "kWarpDevAMG8834Enable" better?

File: src/boot/ksdk1.1.0/CMakeLists.txt [hunk 30]
```diff
@@ -1,142 +1,174 @@
...
-
-# MAP FILE
-SET(CMAKE_EXE_LINKER_FLAGS_DEBUG "${CMAKE_EXE_LINKER_FLAGS_DEBUG}  -Xlinker -Map=debug/Warp.map")
-SET(CMAKE_EXE_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE}  -Xlinker -Map=release/Warp.map")
-
-# BIN AND HEX
-ADD_CUSTOM_COMMAND(TARGET Warp POST_BUILD COMMAND ${CMAKE_OBJCOPY} -Oihex ${EXECUTABLE_OUTPUT_PATH}/Warp.elf ${EXECUTABLE_OUTPUT_PATH}/Warp.hex)
-ADD_CUSTOM_COMMAND(TARGET Warp POST_BUILD COMMAND ${CMAKE_OBJCOPY} -Obinary ${EXECUTABLE_OUTPUT_PATH}/Warp.elf ${EXECUTABLE_OUTPUT_PATH}/Warp.bin)
-ADD_CUSTOM_COMMAND(TARGET Warp POST_BUILD COMMAND ${CMAKE_OBJCOPY} -Osrec ${EXECUTABLE_OUTPUT_PATH}/Warp.elf ${EXECUTABLE_OUTPUT_PATH}/Warp.srec)
+INCLUDE(CMakeForceCompiler)
```
[the remark this answers is not in the document]
Comment (chaturatbs) [reply src/boot/ksdk1.1.0/CMakeLists.txt]: I'm not really sure why its showing diffs when there are changes. I cant figure out how to stop git from showing diffs. Should I revert this file back to the version on master and re-make the changes, manually?

File: src/boot/ksdk1.1.0/devAMG8834.h (hunk 2 above)
Comment (chaturatbs) [reply src/boot/ksdk1.1.0/devAMG8834.h]: Ive renamed all the #defines to "kWarp_devAMG8834_ENABLE" if you prefer otherwise, I'll change it as required

File: src/boot/ksdk1.1.0/gpio_pins.h (hunk 27 above)
Comment (chaturatbs) [reply src/boot/ksdk1.1.0/gpio_pins.h]: I think I've made it worse. now there are diffs everywhere :(

File: src/boot/ksdk1.1.0/CMakeLists.txt (hunk 30 above)
[the remark this answers is not in the document]
Comment (chaturatbs) [reply src/boot/ksdk1.1.0/CMakeLists.txt]: I tried copying all the text from the version on physcomp/master and making the changes but its still showing diffs :(

File: src/boot/ksdk1.1.0/devCC2564C.c [hunk 31]
```diff
@@ -0,0 +1,303 @@
+/*
```
[the remark this answers is not in the document]
Comment (chaturatbs) [reply src/boot/ksdk1.1.0/devCC2564C.c]: I have removed it since its a part of the BT implimentation. I'll have to figure out how to do it properly when i finalise the BTstack for warp

File: src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c [hunk 32]
```diff
@@ -1,2532 +1,3227 @@
...
-					NULL,				/*	spiDeviceState			*/
-					baseAddress,			/*	baseAddress			*/
-					0x00,				/*	minAddress			*/
-					0x2B,				/*	maxAddress			*/
-					repetitionsPerAddress,		/*	repetitionsPerAddress		*/
-					chunkReadsPerAddress,		/*	chunkReadsPerAddress		*/
-					spinDelay,			/*	spinDelay			*/
-					autoIncrement,			/*	autoIncrement			*/
-					sssupplyMillivolts,		/*	sssupplyMillivolts		*/
-
```
[the remark this answers is not in the document]
Comment (chaturatbs) [reply src/boot/ksdk1.1.0/warp-kl03-ksdk1.1-boot.c]: done. :)
